File: Data Structures & Algorithms/products-of-array-discluding-self/submission-1.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int l = nums.size();
        vector<int> res(l, 1);

        for(int i = 1; i < l; i++)
        {
            res[i] = res[i-1] * nums[i-1];
        }

        int postFix = 1;
        for(int j = l-1; j >= 0; j--)
        {
            res[j] *= postFix;
            postFix *= nums[j];
        }

        return res;
    }
};

File: Data Structures & Algorithms/maximum-product-subarray/submission-1.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        n = len(nums)
        res = nums[0]
        pref = 0
        suff = 0

        for i in range(n):
            pref = nums[i] * (pref or 1)
            suff = nums[n - i - 1] * (suff or 1)
            res = max(res, max(pref, suff))
        
        return res